File: app/src/main/java/com/example/firefighters/repositories/WaterPointRepository.java
package com.example.firefighters.repositories;

import android.app.Activity;
import android.widget.Toast;

import com.example.firefighters.models.WaterPointModel;
import com.example.firefighters.models.WaterPointModel;
import com.example.firefighters.tools.ConstantsValues;
import com.example.firefighters.tools.FirebaseManager;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

public class WaterPointRepository {

    private static WaterPointRepository instance;

    public static WaterPointRepository getInstance() {
        if (instance == null)
            instance = new WaterPointRepository();
        return instance;
    }

    public LiveData<QuerySnapshot> getAllWaterPointsQuery() {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        data.setValue(task.getResult());
                    }
                });
        return data;
    }

    public LiveData<QuerySnapshot> getWaterPointsQuery(DocumentSnapshot lastDocument, int limitCount) {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        if (lastDocument != null){
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                    .startAfter(lastDocument)
                    .limit(limitCount)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            data.setValue(task.getResult());
                        }
                    });
        }else{
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                    .limit(limitCount)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            data.setValue(task.getResult());
                        }
                    });
        }
        return data;
    }
    public LiveData<QuerySnapshot> getWaterPointsQuerySnapshot() {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
                    @Override
                    public void onEvent(@Nullable @org.jetbrains.annotations.Nullable QuerySnapshot value, @Nullable @org.jetbrains.annotations.Nullable FirebaseFirestoreException error) {
                        data.setValue(value);
                    }
                });
        return data;
    }
    public LiveData<Integer> saveWaterPoint(WaterPointModel waterPointModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                .limit(1)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            long lastId = 0;
                            if (task.getResult().size() > 0) {
                                WaterPointModel wp = task.getResult().getDocuments().get(0).toObject(WaterPointModel.class);
                                if (wp != null)
                                    lastId = wp.getId();
                            }
                            waterPointModel.setId(lastId+1);
                            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                                    .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                                    .add(waterPointModel)
                                    .addOnCompleteListener(new OnCompleteListener<DocumentReference>() {
                                        @Override
                                        public void onComplete(@NonNull @NotNull Task<DocumentReference> task) {
                                            if (task.isSuccessful()){
                                                data.setValue(1);
                                            }else{
                                                data.setValue(-1);
                                            }
                                        }
                                    });
                        }else {
                            data.setValue(-1);
                        }
                    }
                });

        return data;
    }
    public LiveData<Integer> deleteWaterPoint(WaterPointModel waterPointModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                .whereEqualTo("id", waterPointModel.getId())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            for (DocumentSnapshot document: task.getResult()) {
                                document.getReference()
                                        .delete()
                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                            @Override
                                            public void onComplete(@NonNull @NotNull Task<Void> task) {
                                                if(task.isSuccessful()){
                                                    data.setValue(1);
                                                }else{
                                                    data.setValue(-1);
                                                }
                                            }
                                        });
                            }
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }
    public LiveData<Integer> updateWaterPoint(WaterPointModel waterPointModel){
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.WATER_POINTS_COLLECTION)
                .whereEqualTo("id", waterPointModel.getId())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            for (DocumentSnapshot document: task.getResult()) {
                                document.getReference()
                                        .update("senderMail", waterPointModel.getSenderMail(),
                                                "estimatedQte", waterPointModel.getEstimatedQte(),
                                                "longitude", waterPointModel.getLongitude(),
                                                "latitude", waterPointModel.getLatitude(),
                                                "sourceType", waterPointModel.getSourceType()
                                        )
                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                            @Override
                                            public void onComplete(@NonNull @NotNull Task<Void> task) {
                                                if(task.isSuccessful()){
                                                    data.setValue(1);
                                                }else{
                                                    data.setValue(-1);
                                                }
                                            }
                                        });
                            }
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }

}

File: app/src/main/java/com/example/firefighters/repositories/WaterPointRepository.kt
package com.example.firefighters.repositories

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.firefighters.models.WaterPointModel
import com.example.firefighters.utils.ConstantsValues
import com.example.firefighters.utils.FirebaseUtils
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot

class WaterPointRepository {
    val allWaterPointsQuery: LiveData<QuerySnapshot>
        get() {
            val data = MutableLiveData<QuerySnapshot>()
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
            return data
        }

    fun getWaterPointsQuery(
        lastDocument: DocumentSnapshot?,
        limitCount: Int
    ): LiveData<QuerySnapshot> {
        val data = MutableLiveData<QuerySnapshot>()
        if (lastDocument != null) {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
                ?.startAfter(lastDocument)
                ?.limit(limitCount.toLong())
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
        } else {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
                ?.limit(limitCount.toLong())
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
        }
        return data
    }

    val waterPointsQuerySnapshot: LiveData<QuerySnapshot?>
        get() {
            val data = MutableLiveData<QuerySnapshot?>()
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
                ?.addSnapshotListener { value, _ -> data.value = value }
            return data
        }

    fun saveWaterPoint(waterPointModel: WaterPointModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
            ?.orderBy("id", Query.Direction.DESCENDING)
            ?.limit(1)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    var lastId: Long = 0
                    if ((task.result?.size() ?: 0) > 0) {
                        val wp = task.result?.documents?.get(0)?.toObject(
                            WaterPointModel::class.java
                        )
                        if (wp != null) lastId = wp.id
                    }
                    waterPointModel.id = lastId + 1
                    FirebaseUtils.instance?.firebaseFirestoreInstance
                        ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
                        ?.add(waterPointModel)
                        ?.addOnCompleteListener {
                            if (it.isSuccessful) {
                                data.setValue(1)
                            } else {
                                data.setValue(-1)
                            }
                        }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun deleteWaterPoint(waterPointModel: WaterPointModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
            ?.whereEqualTo("id", waterPointModel.id)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in task.result ?: listOf()) {
                        document.reference
                            .delete()
                            .addOnCompleteListener {
                                if (it.isSuccessful) {
                                    data.setValue(1)
                                } else {
                                    data.setValue(-1)
                                }
                            }
                    }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun updateWaterPoint(waterPointModel: WaterPointModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
            ?.whereEqualTo("id", waterPointModel.id)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in task.result ?: listOf()) {
                        document.reference
                            .update(
                                "senderMail", waterPointModel.senderMail,
                                "estimatedQte", waterPointModel.estimatedQte,
                                "longitude", waterPointModel.longitude,
                                "latitude", waterPointModel.latitude,
                                "sourceType", waterPointModel.sourceType
                            )
                            .addOnCompleteListener {
                                if (it.isSuccessful) {
                                    data.setValue(1)
                                } else {
                                    data.setValue(-1)
                                }
                            }
                    }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun getWaterPointModel(id: Int): LiveData<WaterPointModel?> {
        val data = MutableLiveData<WaterPointModel?>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.WATER_POINTS_COLLECTION)
            ?.whereEqualTo("id", id)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    if ((task.result?.size() ?: 0) > 0) {
                        val waterPointModel = task.result?.documents?.get(0)?.toObject(
                            WaterPointModel::class.java
                        )
                        data.setValue(waterPointModel)
                    } else {
                        data.setValue(null)
                    }
                } else {
                    data.setValue(null)
                }
            }
        return data
    }
}
